diff --git a/src/go/build/vendor_test.go b/src/go/build/vendor_test.go
index 7f6237ffd5..fe552e5326 100644
--- a/src/go/build/vendor_test.go
+++ b/src/go/build/vendor_test.go
@@ -22,6 +22,7 @@ var allowedPackagePrefixes = []string{
 	"github.com/google/pprof",
 	"github.com/ianlancetaylor/demangle",
 	"rsc.io/markdown",
+	"github.com/golang-fips/openssl/v2",
 }
 
 // Verify that the vendor directories contain only packages matching the list above.
