diff --git a/src/crypto/tls/handshake_client.go b/src/crypto/tls/handshake_client.go
index e4ef45b882..9e2549847a 100644
--- a/src/crypto/tls/handshake_client.go
+++ b/src/crypto/tls/handshake_client.go
@@ -149,7 +149,11 @@ func (c *Conn) makeClientHello() (*clientHelloMsg, *keySharePrivateKeys, *echCon
 			hello.cipherSuites = append(hello.cipherSuites, defaultCipherSuitesTLS13NoAES...)
 		}
 
-		curveID := config.curvePreferences(maxVersion)[0]
+    curvePreferences := config.curvePreferences(maxVersion)
+    if len(curvePreferences) == 0 {
+                  return nil, nil, nil, errors.New("tls: No allowed curves configured")
+    }
+    curveID := curvePreferences[0]
 		keyShareKeys = &keySharePrivateKeys{curveID: curveID}
 		if curveID == x25519Kyber768Draft00 {
 			keyShareKeys.ecdhe, err = generateECDHEKey(config.rand(), X25519)
